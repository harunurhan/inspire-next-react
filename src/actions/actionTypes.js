export const SEARCHING = 'SEARCHING';
export const SEARCH_SUCCESS = 'SEARCH_SUCCESS';
export const SEARCH_ERROR = 'SEARCH_ERROR';
