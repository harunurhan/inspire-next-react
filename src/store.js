import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import createHistory from 'history/createBrowserHistory';
import { routerMiddleware } from 'react-router-redux';
/* eslint-disable import/no-extraneous-dependencies */
import { createLogger } from 'redux-logger';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
/* eslint-disable import/no-extraneous-dependencies */

import reducers from './reducers';
import http from './common/http';
import queryParamsParser from './middlewares/queryParamsParser';

export const thunkMiddleware = thunk.withExtraArgument(http);

export const history = createHistory();
const reduxRouterMiddleware = routerMiddleware(history);

const queryParamsParserMiddleware = queryParamsParser();

const getMiddleware = () => {
  if (process.env.NODE_ENV === 'production') {
    return applyMiddleware(
      reduxRouterMiddleware,
      queryParamsParserMiddleware,
      thunkMiddleware,
    );
  } else {
    return applyMiddleware(
      createLogger(),
      reduxRouterMiddleware,
      queryParamsParserMiddleware,
      thunkMiddleware,
    );
  }
};

export const store = createStore(reducers, composeWithDevTools(getMiddleware()));
