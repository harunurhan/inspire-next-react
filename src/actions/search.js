import { push } from 'react-router-redux';
import { stringify } from 'qs';

import * as types from './actionTypes';

function searching() {
  return {
    type: types.SEARCHING,
  };
}

function searchSuccess(result) {
  return {
    type: types.SEARCH_SUCCESS,
    payload: result,
  };
}

function searchError(error) {
  return {
    type: types.SEARCH_ERROR,
    payload: error,
  };
}

function getSearchUrl(state, query) {
  const locationQuery = state.router.location.query;
  const baseQuery = state.search.getIn(['scope', 'query']).toJS();
  const pathname = state.search.getIn(['scope', 'pathname']);
  const newQuery = {
    ...locationQuery,
    ...baseQuery,
    ...query,
  };
  const queryString = stringify(newQuery);
  return `${pathname}?${queryString}`;
}

export default function search(query) {
  return async (dispatch, getState, http) => {
    dispatch(searching());

    const state = getState();
    const url = getSearchUrl(state, query);
    if (query) {
      dispatch(push(url));
    }
    try {
      const response = await http.get(url);

      dispatch(searchSuccess(response.data));
    } catch (error) {
      dispatch(searchError(error.data));
    }
  };
}
