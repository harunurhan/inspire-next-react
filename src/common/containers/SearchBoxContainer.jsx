import { connect } from 'react-redux';

import toJS from '../immutableToJS';
import SearchBox from '../components/SearchBox';
import search from '../../actions/search';

const stateToProps = state => ({
  defaultValue: state.router.location.query.q,
});

export const dispatchToProps = dispatch => ({
  onSearch(value) {
    dispatch(search({ q: value }));
  },
});

export default connect(stateToProps, dispatchToProps)(toJS(SearchBox));
