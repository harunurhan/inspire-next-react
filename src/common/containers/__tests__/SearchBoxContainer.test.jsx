import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import { getStoreWithState, getStore } from '../../../fixtures/store';
import { SEARCHING } from '../../../actions/actionTypes';
import SearchBoxContainer, { dispatchToProps } from '../SearchBoxContainer';

describe('SearchBoxContainer', () => {
  it('renders initial state with initial url query q param', () => {
    const store = getStoreWithState({
      router: { location: { query: { q: 'test' } } },
    });
    const wrapper = mount((
      <Provider store={store}>
        <SearchBoxContainer />
      </Provider>
    ));
    expect(wrapper).toMatchSnapshot();
  });

  it('dispatches search onSearch', () => {
    const store = getStore();
    const props = dispatchToProps(store.dispatch);
    props.onSearch();
    const actions = store.getActions();
    expect(actions.some(action => action.type === SEARCHING)).toBe(true);
  });
});
