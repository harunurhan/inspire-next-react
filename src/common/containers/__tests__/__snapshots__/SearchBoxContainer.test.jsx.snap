// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBoxContainer renders initial state with initial url query q param 1`] = `
<div>
  <span
    className="ant-input-search ant-input-search-enter-button ant-input-affix-wrapper"
  >
    <input
      className="ant-input"
      defaultValue="test"
      disabled={false}
      onKeyDown={[Function]}
      placeholder={null}
      style={null}
      type="text"
    />
    <span
      className="ant-input-suffix"
    >
      <button
        className="ant-btn ant-input-search-button ant-btn-primary"
        onClick={[Function]}
        type="button"
      >
        <i
          className="anticon anticon-search"
        />
      </button>
    </span>
  </span>
</div>
`;
